src/rules/BanFormattedCompMessage.js

```diff
@@ -0,0 +1,125 @@
+        if (ir.type !== this.type) {
+            throw new Error(`Unexpected representation type!`);
+        }
+
+        const /** @type {unknown} */ tree = ir.getRepresentation();
+
+        // find JSX opening elements with matching name
+        const jsxElements = /** @type {unknown[]} */ (
+            jsonpath.query(
+                tree,
+                "$..[?(@.type == 'JSXOpeningElement' && @.name.name == 'FormattedCompMessage')].name"
+            )
+        );
+
+        // find every non-JSX identifier with matching name (catch all kinds of imports and requires);
+        // avoid duplicate matches (e.g. an import statement outputs two identifiers in the same range)
+        const identifiers = distinct(
+            /** @type {unknown[]} */ (
+                jsonpath.query(
+                    tree,
+                    "$..[?(@.type == 'Identifier' && @.name == 'FormattedCompMessage')]"
+                )
+            ),
+            (one, other) => rangesEqual(getRange(one), getRange(other))
+        );
+
+        return [...jsxElements, ...identifiers]
+            .map((node) => getRange(node))
+            .sort((a, b) => rangesCompare(a, b))
+            .map((range) => {
+                return new Result({
+                    pathName: ir.filePath,
+                    severity: "error",
+                    rule: this,
+                    description: `Do not use deprecated FormattedCompMessage component.`,
+                    id: undefined,
+                    // @TODO make a real highlight once IR contains raw content of the linted source file
+                    highlight: `Range: <e0>${range[0]}:${range[1]}</e0>`
+                });
+            });
+    }
+}
+
+/** @typedef {{ range: [number, number] }} FlowRange */
+const isFlowRange = /** @type {(node: any) => node is FlowRange} */ (node) => {
+    return (
+        "range" in node &&
+        Array.isArray(node.range) &&
+        node.range.length === 2 &&
+        !node.range.some((/** @type {any} */ idx) => "number" !== typeof idx)
+    );
+};
+
+/** @typedef {{ start: number; end: number }} AcornRange */
+const isAcornRange = /** @type {(node: any) => node is AcornRange} */ (
+    node
+) => {
+    return "number" === typeof node.start && "number" === typeof node.end;
+};
+
+/** @typedef {[number, number]} Range */
+const getRange = /** @type {(node: any) => Range} */ (node) => {
+    if (isFlowRange(node)) {
+        return node.range;
+    }
+    if (isAcornRange(node)) {
+        return [node.start, node.end];
+    }
+    throw new Error(`Unexpected AST node shape: missing range`);
+};
+
+const rangesEqual = (/** @type {Range} */ one, /** @type {Range} */ other) => {
+    return rangesCompare(one, other) === 0;
+};
+
+const rangesCompare = (
+    /** @type {Range} */ one,
+    /** @type {Range} */ other
+) => {
+    return one[0] === other[0] ? one[0] - other[0] : one[1] - other[1];
```

Comment:
If the first elements are equal, then return the difference between them -- which will always be zero. Maybe this function isn't doing the right thing? I think you meant:

```
return one[0] === other[0] ? one[1] - other[1] : one[0] - other[0];
```